File: java/src/test/java/com/threedimensionalloadingcvrp/validator/routing/DispatchedItemsTest.java
package com.threedimensionalloadingcvrp.validator.routing;

import com.threedimensionalloadingcvrp.validator.model.*;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static com.threedimensionalloadingcvrp.validator.constraints.Routing.checkDispatchedItems;

public class DispatchedItemsTest {
    @Test
    public void Demand2NotFulfilled() {
        ItemType type1 = new ItemType(1, 0, 0, 0, 0, false, 0);
        ItemType type2 = new ItemType(2, 0, 0, 0, 0, false, 0);

        Item item1 = new Item(0, 1, Rotation.none, type1);
        Item item2 = new Item(1, 1, Rotation.none, type1);
        Item item3 = new Item(2, 1, Rotation.none, type1);

        Item item4 = new Item(3, 1, Rotation.none, type2);
        Item item5 = new Item(4, 1, Rotation.none, type2);

        // Instance Creation
        Customer depot = new Customer(0, 0, 0, 0, 0, 0, 0, 0, 0);
        Customer customer = new Customer(1, 0, 0, 5, 0, 0, 0, 0, 0);
        Instance instance = new Instance("", null, null, Arrays.asList(depot, customer), 0, false, Arrays.asList(item1, item2, item3, item4, item5));
        Demand demand1 = new Demand(1, 3);
        Demand demand2 = new Demand(2, 3);
        instance.getCustomers().get(1).getDemands().add(demand1);
        instance.getCustomers().get(1).getDemands().add(demand2);

        // Tour Creation
        Tour tour = new Tour(1, Collections.singletonList(1), Arrays.asList(0, 1, 2, 3, 4));

        // Solution Creation
        Solution solution = new Solution(Collections.singletonList(tour));

        // Test
        Assertions.assertFalse(checkDispatchedItems(solution, instance, false));
    }

    @Test
    public void Demand2Split() {
        ItemType type1 = new ItemType(1, 0, 0, 0, 0, false, 0);
        ItemType type2 = new ItemType(2, 0, 0, 0, 0, false, 0);

        Item item1 = new Item(0, 1, Rotation.none, type1);
        Item item2 = new Item(1, 1, Rotation.none, type1);
        Item item3 = new Item(2, 1, Rotation.none, type1);
        Item item4 = new Item(3, 1, Rotation.none, type2);

        Item item5 = new Item(4, 2, Rotation.none, type2);
        Item item6 = new Item(5, 2, Rotation.none, type2);

        // Instance Creation
        Customer depot = new Customer(0, 0, 0, 0, 0, 0, 0, 0, 0);
        Customer customer1 = new Customer(1, 0, 0, 4, 0, 0, 0, 0, 0);
        Customer customer2 = new Customer(2, 0, 0, 2, 0, 0, 0, 0, 0);
        Instance instance = new Instance("", null, null, Arrays.asList(depot, customer1, customer2), 0, false, Arrays.asList(item1, item2, item3, item4, item5, item6));
        Demand demand1 = new Demand(1, 3);
        Demand demand2 = new Demand(2, 1);
        Demand demand3 = new Demand(2, 2);
        instance.getCustomers().get(1).getDemands().add(demand1);
        instance.getCustomers().get(1).getDemands().add(demand2);
        instance.getCustomers().get(2).getDemands().add(demand3);

        // Tour Creation
        Tour tour1 = new Tour(1, List.of(1, 2), Arrays.asList(0, 1, 2, 4, 5));
        Tour tour2 = new Tour(2, List.of(1), Arrays.asList(3));

        // Solution Creation
        Solution solution = new Solution(List.of(tour1, tour2));

        // Test
        Assertions.assertTrue(checkDispatchedItems(solution, instance, false));
    }

    @Test
    public void DemandFulfilled() {
        ItemType type1 = new ItemType(1, 0, 0, 0, 0, false, 0);
        ItemType type2 = new ItemType(2, 0, 0, 0, 0, false, 0);

        Item item1 = new Item(0, 1, Rotation.none, type1);
        Item item2 = new Item(1, 1, Rotation.none, type1);
        Item item3 = new Item(2, 1, Rotation.none, type1);

        Item item4 = new Item(3, 1, Rotation.none, type2);
        Item item5 = new Item(4, 1, Rotation.none, type2);

        // Instance Creation
        Customer depot = new Customer(0, 0, 0, 0, 0, 0, 0, 0, 0);
        Customer customer = new Customer(1, 0, 0, 5, 0, 0, 0, 0, 0);
        Instance instance = new Instance("", null, null, Arrays.asList(depot, customer), 0, false, Arrays.asList(item1, item2, item3, item4, item5));
        Demand demand1 = new Demand(1, 3);
        Demand demand2 = new Demand(2, 2);
        instance.getCustomers().get(1).getDemands().add(demand1);
        instance.getCustomers().get(1).getDemands().add(demand2);

        // Tour Creation
        Tour tour = new Tour(1, Collections.singletonList(1), Arrays.asList(0, 1, 2, 3, 4));

        // Solution Creation
        Solution solution = new Solution(Collections.singletonList(tour));

        // Test
        Assertions.assertTrue(checkDispatchedItems(solution, instance, false));
    }
}

File: cpp/UnitTest/DispatchedItemsTest.cpp
#include "pch.h"
#include "CppUnitTest.h"
#include "../Validator/ConstraintsRouting.h"
#include <string>
#include <vector>

using namespace Microsoft::VisualStudio::CppUnitTestFramework;
using namespace validator;

namespace RoutingConstraints {
	TEST_CLASS(DispatchedItemsTest) {
	public:
		TEST_METHOD(Demand2NotFulfilled) {
			ItemType type1(1, 0, 0, 0, 0, false, 0);
			ItemType type2(2, 0, 0, 0, 0, false, 0);

			Item item1(0, 1, 0, type1);
			Item item2(1, 1, 0, type1);
			Item item3(2, 1, 0, type1);

			Item item4(3, 1, 0, type2);
			Item item5(4, 1, 0, type2);

			// Instance Creation
			Customer depot(0, 0, 0, 0, 0, 0, 0, 0, 0);
			Customer customer(1, 0, 0, 5, 0, 0, 0, 0, 0);
			std::vector<Customer> customers{ depot, customer };
			std::vector<ItemType> itemtypes{ type1, type2 };
			Instance instance("", Vehicle(0, 0, 0, 0, 0, 0, 0, 0, 0), itemtypes, customers, 0, 0);
			instance.items.emplace_back(item1);
			instance.items.emplace_back(item2);
			instance.items.emplace_back(item3);
			instance.items.emplace_back(item4);
			instance.items.emplace_back(item5);
			std::pair<int, int> demand1(1, 3);
			std::pair<int, int> demand2(2, 3);
			instance.customers.at(1).demands.emplace_back(demand1);
			instance.customers.at(1).demands.emplace_back(demand2);

			// Tour Creation
			std::vector<unsigned int> customer_ids{ 1 };
			std::vector<unsigned int> item_ids{ 0, 1, 2, 3, 4  };
			Tour tour(1, customer_ids, item_ids);

			// Solution Creation
			std::vector<Tour> tours{ tour };
			Solution solution(tours);

			// Test
			Assert::AreEqual(false, ConstraintsRouting::checkDispatchedItems(solution, instance, false));
		}

		TEST_METHOD(Demand2Split) {
			ItemType type1(1, 0, 0, 0, 0, false, 0);
			ItemType type2(2, 0, 0, 0, 0, false, 0);

			Item item1(0, 1, 0, type1);
			Item item2(1, 1, 0, type1);
			Item item3(2, 1, 0, type1);
			Item item4(3, 1, 0, type2);

			Item item5(4, 2, 0, type2);
			Item item6(5, 2, 0, type2);

			// Instance Creation
			Customer depot(0, 0, 0, 0, 0, 0, 0, 0, 0);
			Customer customer1(1, 0, 0, 4, 0, 0, 0, 0, 0);
			Customer customer2(2, 0, 0, 2, 0, 0, 0, 0, 0);
			std::vector<Customer> customers{ depot, customer1, customer2 };
			std::vector<ItemType> itemtypes{ type1, type2 };
			Instance instance("", Vehicle(0, 0, 0, 0, 0, 0, 0, 0, 0), itemtypes, customers, 0, 0);
			instance.items.emplace_back(item1);
			instance.items.emplace_back(item2);
			instance.items.emplace_back(item3);
			instance.items.emplace_back(item4);
			instance.items.emplace_back(item5);
			instance.items.emplace_back(item6);
			std::pair<int, int> demand1(1, 3);
			std::pair<int, int> demand2(2, 1);
			std::pair<int, int> demand3(2, 2);
			instance.customers.at(1).demands.emplace_back(demand1);
			instance.customers.at(1).demands.emplace_back(demand2);
			instance.customers.at(2).demands.emplace_back(demand3);

			// Tour Creation
			std::vector<unsigned int> customer_ids1{ 1, 2 };
			std::vector<unsigned int> customer_ids2{ 1 };
			std::vector<unsigned int> item_ids1{ 0, 1, 2, 4, 5 };
			std::vector<unsigned int> item_ids2{ 3 };

			Tour tour1(1, customer_ids1, item_ids1);
			Tour tour2(1, customer_ids2, item_ids2);

			// Solution Creation
			std::vector<Tour> tours{ tour1, tour2 };
			Solution solution(tours);

			// Test
			Assert::AreEqual(true, ConstraintsRouting::checkDispatchedItems(solution, instance, false));
		}

		TEST_METHOD(DemandFulfilled) {
			ItemType type1(1, 0, 0, 0, 0, false, 0);
			ItemType type2(2, 0, 0, 0, 0, false, 0);

			Item item1(0, 1, 0, type1);
			Item item2(1, 1, 0, type1);
			Item item3(2, 1, 0, type1);

			Item item4(3, 1, 0, type2);
			Item item5(4, 1, 0, type2);

			// Instance Creation
			Customer depot(0, 0, 0, 0, 0, 0, 0, 0, 0);
			Customer customer(1, 0, 0, 5, 0, 0, 0, 0, 0);
			std::vector<Customer> customers{ depot, customer };
			std::vector<ItemType> itemtypes{ type1, type2 };
			Instance instance("", Vehicle(0, 0, 0, 0, 0, 0, 0, 0, 0), itemtypes, customers, 0, 0);
			instance.items.emplace_back(item1);
			instance.items.emplace_back(item2);
			instance.items.emplace_back(item3);
			instance.items.emplace_back(item4);
			instance.items.emplace_back(item5);
			std::pair<int, int> demand1(1, 3);
			std::pair<int, int> demand2(2, 2);
			instance.customers.at(1).demands.emplace_back(demand1);
			instance.customers.at(1).demands.emplace_back(demand2);

			// Tour Creation
			std::vector<unsigned int> customer_ids{ 1 };
			std::vector<unsigned int> item_ids{ 0, 1, 2, 3, 4 };
			Tour tour(1, customer_ids, item_ids);

			// Solution Creation
			std::vector<Tour> tours{ tour };
			Solution solution(tours);

			// Test
			Assert::AreEqual(true, ConstraintsRouting::checkDispatchedItems(solution, instance, false));
		}
	};
}